crates/gitbutler-repo/src/rebase.rs

```diff
@@ -48,54 +52,224 @@ pub fn cherry_rebase_group(
+                    if !ctx.project().succeeding_rebases {
+                        return Err(anyhow!("failed to rebase")).context(Marker::BranchConflict);
+                    }
+                    commit_conflicted_cherry_result(ctx, head, to_rebase, cherrypick_index)
+                } else {
+                    commit_unconflicted_cherry_result(ctx, head, to_rebase, cherrypick_index)
                 }
-
-                let merge_tree_oid = cherrypick_index
-                    .write_tree_to(ctx.repository())
-                    .context("failed to write merge tree")?;
-
-                let merge_tree = ctx
-                    .repository()
-                    .find_tree(merge_tree_oid)
-                    .context("failed to find merge tree")?;
-
-                let commit_headers = to_rebase.gitbutler_headers();
-
-                let commit_oid = ctx
-                    .repository()
-                    .commit_with_signature(
-                        None,
-                        &to_rebase.author(),
-                        &to_rebase.committer(),
-                        &to_rebase.message_bstr().to_str_lossy(),
-                        &merge_tree,
-                        &[&head],
-                        commit_headers,
-                    )
-                    .context("failed to create commit")?;
-
-                ctx.repository()
-                    .find_commit(commit_oid)
-                    .context("failed to find commit")
             },
         )?
         .id();
 
     Ok(new_head_id)
 }
+
+fn commit_unconflicted_cherry_result<'repo>(
+    ctx: &'repo CommandContext,
+    head: git2::Commit<'repo>,
+    to_rebase: git2::Commit,
+    mut cherrypick_index: git2::Index,
+) -> Result<git2::Commit<'repo>> {
+    let repository = ctx.repository();
+    let commit_headers = to_rebase.gitbutler_headers();
+
+    let is_merge_commit = to_rebase.parent_count() > 0;
+
+    let merge_tree_oid = cherrypick_index
+        .write_tree_to(repository)
+        .context("failed to write merge tree")?;
+
+    // Remove empty merge commits
+    if is_merge_commit && merge_tree_oid == head.tree_id() {
+        return Ok(head);
+    }
+
+    let merge_tree = repository
+        .find_tree(merge_tree_oid)
+        .context("failed to find merge tree")?;
+
+    let commit_headers = commit_headers.map(|commit_headers| CommitHeadersV2 {
+        conflicted: None,
+        ..commit_headers
+    });
+
+    let commit_oid = repository
+        .commit_with_signature(
+            None,
+            &to_rebase.author(),
+            &to_rebase.committer(),
+            &to_rebase.message_bstr().to_str_lossy(),
+            &merge_tree,
+            &[&head],
+            commit_headers,
+        )
+        .context("failed to create commit")?;
+
+    repository
+        .find_commit(commit_oid)
+        .context("failed to find commit")
+}
+
+fn commit_conflicted_cherry_result<'l>(
```

Comment:
btw why not the conventional life time names (ie. 'a 'b)?

Reply:
:+1:

Reply:
You're right. That was a very short and meaningless name.

```rs
fn commit_unconflicted_cherry_result<'repository>(
    ctx: &'repository CommandContext,
    head: git2::Commit<'repository>,
    to_rebase: git2::Commit,
    mut cherrypick_index: git2::Index,
) -> Result<git2::Commit<'repository>> {
```